### PR TITLE
Refactor: Files except sonar-related files

### LINKED ISSUE #36: [][FE][REFACTOR] useEffect optimization

### PR DESCRIPTION
**Pull request by hellenGUAN**
# Description

- simplify functions in useEffect()
- rewrite call API functions 
- separate calculation to several functions
- return response instead of setting value in get functions

### CODE REVIEW

File: src/App/component/CommitsPage.jsx [hunk 1]
```diff
@@ -174,58 +162,56 @@ function CommitsPage(prop) {
...
       }
     }
   }, [isLoading]);
 
-  useEffect(() => {
-    const { startMonth, endMonth } = prop
-
-    let chartDataset = { labels: [], data: { team: [] } }
+  const setCommitChart = () => {
+    let dataset = { labels: [], data: { team: [] } }
```
Comment (jn2657) [inline src/App/component/CommitsPage.jsx]:
```suggestion
    const dataset = { labels: [], data: { team: [] } }
```

File: src/App/component/CommitsPage.jsx [hunk 2]
```diff
@@ -64,96 +65,84 @@ function CommitsPage(prop) {
...
+      console.error(e)
     }
   }
 
-  const getCommitFromGitLab = () => {
+  useEffect(() => {
+    loadInitialProjectInfo()
+  }, [])
+
+  const getCommit = async () => {
```
Comment (jn2657) [inline src/App/component/CommitsPage.jsx]:
```suggestion
  const updateCommit = async () => {
```

File: src/App/component/CommitsPage.jsx [hunk 3]
```diff
@@ -64,96 +65,84 @@ function CommitsPage(prop) {
...
+      const query = repo.url.split(repo.type + ".com/")[1]
+      const repoOwner = query.split("/")[0]
+      const repoName = query.split("/")[1]
+
+      try {
+        const response = await sendPVSBackendRequest('GET', `/${repo.type}/commits/${repoOwner}/${repoName}`)
+        setCommitListData(response)
+      } catch (e) {
+        alert(e.response?.status)
+        console.error(e)
```
Comment (jn2657) [inline src/App/component/CommitsPage.jsx]:
```suggestion
        console.error(e)
        loadingCommitsEnd()
```

File: src/App/component/CommitsPage.jsx [hunk 4]
```diff
@@ -64,96 +65,84 @@ function CommitsPage(prop) {
...
+      }
```
Comment (jn2657) [inline src/App/component/CommitsPage.jsx]:
```suggestion
      }
      loadingCommitsEnd()
```

File: src/App/component/CommitsPage.jsx [hunk 5]
```diff
@@ -162,10 +151,9 @@ function CommitsPage(prop) {
   // Default get commits from database
   useEffect(() => {
     if (Object.keys(currentProject).length !== 0) {
-      handleToggle()
-      getGitHubCommitFromDB()
-      getGitLabCommitFromDB()
-      handleClose()
+      loadingCommitsStart()
+      getCommitFromDB()
+      loadingCommitsEnd()
```
Comment (jn2657) [inline src/App/component/CommitsPage.jsx]:
```suggestion
```

File: src/App/component/ComparisonPage.jsx [hunk 6]
```diff
@@ -220,14 +209,40 @@ function ComparisonPage(prop) {
...
+          else if (branch === rightBranchSelected) {
+            dataset.data[branch].push(commitListDataRight.filter(commit => {
+              return moment(commit.committedDate).format("YYYY-MM") === month.format("YYYY-MM")
+            }).length)
+          }
+        }
+      }
+    }
+    setDataForTeamCommitChart(dataset)
+  }
```
Comment (jn2657) [inline src/App/component/ComparisonPage.jsx]:
```suggestion
  const setComparisonChart = () => {
    const { startMonth, endMonth } = prop
    const chartDataset = { labels: [], data: {} }
    new Set(selectedBranchList).forEach(branch => {
      chartDataset.data[branch] = []
    })
    for (let month = moment(startMonth); month <= moment(endMonth); month = month.add(1, 'months')) {
      chartDataset.labels.push(month.format("YYYY-MM"))
      for (const branch in chartDataset.data) {
        chartDataset.data[branch].push(getCommitCountFromSelectedBranch(branch, month))
      }
    }
    setDataForTeamCommitChart(chartDataset)
  }

  const getCommitCountFromSelectedBranch = (branch, month) => {
    if (branch !== "" && branch == leftBranchSelected) {
      return commitListDataLeft.filter(commit => {
        return moment(commit.committedDate).format("YYYY-MM") === month.format("YYYY-MM")
      }).length
    }

    if (branch !== "" && branch == rightBranchSelected) {
      return commitListDataRight.filter(commit => {
        return moment(commit.committedDate).format("YYYY-MM") === month.format("YYYY-MM")
      }).length
    }
  }
```

File: src/App/component/ContributionPage.jsx [hunk 7]
```diff
@@ -61,94 +61,85 @@ function ContributionPage(prop) {
...
+      const repoName = query.split("/")[1]
+
+      try {
+        const response = await sendPVSBackendRequest('GET', `/${repo.type}/commits/${repoOwner}/${repoName}`)
+        setCommitListData(response)
+      } catch (e) {
+        alert(e.response?.status)
+        console.error(e)
+      }
     }
```
Comment (jn2657) [inline src/App/component/ContributionPage.jsx]:
```suggestion
      } catch (e) {
        alert(e.response?.status)
        console.error(e)
        loadingCommitsEnd()
      }
      loadingCommitsEnd()
    }
```

File: src/App/component/ContributionPage.jsx [hunk 8]
```diff
@@ -157,10 +148,9 @@ function ContributionPage(prop) {
   // Default get commits from database
   useEffect(() => {
     if (Object.keys(currentProject).length !== 0) {
-      handleToggle()
-      getGitHubCommitFromDB()
-      getGitLabCommitFromDB()
-      handleClose()
+      loadingCommitsStart()
+      getCommitFromDB()
+      loadingCommitsEnd()
```
Comment (jn2657) [inline src/App/component/ContributionPage.jsx]:
```suggestion
```

File: src/App/component/CommitsPage.jsx [hunk 9]
```diff
@@ -64,108 +65,96 @@ function CommitsPage(prop) {
...
-    setOpen(false);
-  };
-  const handleToggle = () => {
-    setOpen(!open);
-  };
+  const loadingCommitsEnd = () => {
+    setOpen(false)
+  }
+  const loadingCommitsStart = () => {
+    setOpen(!open)
```
Comment (VincentETN) [inline src/App/component/CommitsPage.jsx]:
```suggestion
    setOpen(true)
```
If the function name has a keyword "Start", suggest setting `open` true directly.

File: src/App/component/ComparisonPage.jsx [hunk 10]
```diff
@@ -80,44 +82,41 @@ function ComparisonPage(prop) {
...
-    setOpen(false);
-  };
-  const handleToggle = () => {
-    setOpen(!open);
-  };
+  const loadingCommitsEnd = () => {
+    setOpen(false)
+  }
+  const loadingCommitsStart = () => {
+    setOpen(!open)
```
Comment (VincentETN) [inline src/App/component/ComparisonPage.jsx]: Ditto.

File: src/App/component/ComparisonPage.jsx [hunk 11]
```diff
@@ -220,14 +209,42 @@ function ComparisonPage(prop) {
...
+    }
+  }
+
   useEffect(() => {
     if (Object.keys(currentProject).length !== 0) {
-      handleToggle()
+      loadingCommitsStart()
       getBranches()
       getCommitFromDBLeft(leftBranchSelected)
       getCommitFromDBRight(rightBranchSelected)
```
Comment (VincentETN) [inline src/App/component/ComparisonPage.jsx]: Function 可以用同一個。
```suggestion
      getCommitFromDB(leftBranchSelected)
      getCommitFromDB(rightBranchSelected)
```

File: src/App/component/ContributionPage.jsx [hunk 12]
```diff
@@ -61,94 +61,87 @@ function ContributionPage(prop) {
...
   const [open, setOpen] = useState(false);
   const [isLoading, setLoading] = useState(false);
-  const handleClose = () => {
+  const loadingCommitsEnd = () => {
     setOpen(false);
-  };
-  const handleToggle = () => {
+  }
+  const loadingCommitsStart = () => {
     setOpen(!open);
```
Comment (VincentETN) [inline src/App/component/ContributionPage.jsx]: 跟前面一樣。

File: src/App/component/AddRepositoryDialog.jsx [hunk 13]
```diff
@@ -37,23 +54,21 @@ export default function AddRepositoryDialog({ open, reloadProjects, handleClose,
...
-              .then(() => {
-                reloadProjects()
-                handleClose()
-              })
-              .catch((e) => {
-                alert(e.response?.status)
-                console.error(e)
-              })
+            try {
+              await Axios.post(`http://localhost:9100/pvs-api/project/${projectId}/repository/${repoType}`, payload, config)
```
Comment (Xanonymous-GitHub) [inline src/App/component/AddRepositoryDialog.jsx]: Suggest replace with `sendPVSBackendRequest` you just declared above.

File: src/App/component/AddRepositoryDialog.jsx [hunk 14]
```diff
@@ -37,23 +54,21 @@ export default function AddRepositoryDialog({ open, reloadProjects, handleClose,
       checker.push(checkRepositoryURL(repositoryURL));
 
       Promise.all(checker)
```
Comment (Xanonymous-GitHub) [inline src/App/component/AddRepositoryDialog.jsx]: Suggest not use both `.then` and `async await`.
```suggestion
      const urlCheckResults = await Promise.all(checker)
```

File: src/App/component/AddRepositoryDialog.jsx [hunk 15]
```diff
@@ -37,23 +54,21 @@ export default function AddRepositoryDialog({ open, reloadProjects, handleClose,
...
-        .then((response) => {
+        .then(async (response) => {
```
Comment (Xanonymous-GitHub) [inline src/App/component/AddRepositoryDialog.jsx]: ditto.
```suggestion
        for (const checlResult of urlCheckResults) {
          ...
        }
```

File: src/App/component/AddRepositoryDialog.jsx (hunk 13 above)
Comment (jn2657) [reply src/App/component/AddRepositoryDialog.jsx]: How to use `sendPVSBackendRequest` to send payload?

File: src/App/component/AddRepositoryDialog.jsx [hunk 16]
```diff
@@ -65,45 +80,41 @@ export default function AddRepositoryDialog({ open, reloadProjects, handleClose,
...
+        return false
       }
     }
 
     if (repoType === "gitlab") {
       try {
-        await Axios.get(`http://localhost:9100/pvs-api/repository/gitlab/check?url=${url}`,
-          { headers: { "Authorization": `${jwtToken}` } });
-        return true;
+        await sendPVSBackendRequest('GET', `/repository/gitlab/check?url=${url}`)
```
Comment (Xanonymous-GitHub) [inline src/App/component/AddRepositoryDialog.jsx]: Suggest use `axios` api to send a GET request with query parameters.

So may be you can change the function to this:

```js
  const sendPVSBackendRequest = async (method, url, params) => {
    const baseURL = 'http://localhost:9100/pvs-api'
    const requestConfig = {
      baseURL,
      url,
      method,
      headers,
      params
    }
    return (await Axios.request(requestConfig))?.data
  }
```

and change this line to

```js
await sendPVSBackendRequest('GET', `/repository/gitlab/check`, { url })
```

File: src/App/component/AddRepositoryDialog.jsx [hunk 17]
```diff
@@ -28,6 +28,23 @@ export default function AddRepositoryDialog({ open, reloadProjects, handleClose,
 
   let repositoryURL = ""
 
+  const config = {
+    headers: {
+      ...(jwtToken && { "Authorization": jwtToken })
+    }
+  }
```
Comment (Xanonymous-GitHub) [inline src/App/component/AddRepositoryDialog.jsx]: A `config` here only includes `headers` may causes some misunderstanding with the type that `axios` has already declared.

Please see [HERE](https://github.com/axios/axios/blob/4fbf61dc751ad69c7df9977d6edf7cf7ee87a5fb/index.d.ts#L76) for more details.

```suggestion
  const headers = { ...(jwtToken && { "Authorization": jwtToken }) }
```

File: src/App/component/CodeBasePage.jsx [hunk 18]
```diff
@@ -64,91 +65,82 @@ function CodeBasePage(prop) {
...
-      .catch((e) => {
-        alert(e.response?.status)
-        console.error(e)
-      })
-  }, [])
+  const config = {
+    headers: {
+      ...(jwtToken && { "Authorization": jwtToken })
+    }
+  }
```
Comment (Xanonymous-GitHub) [inline src/App/component/CodeBasePage.jsx]: ditto.
```suggestion
  const headers = { ...(jwtToken && { "Authorization": jwtToken }) }
```

File: src/App/component/CodeBasePage.jsx [hunk 19]
```diff
@@ -64,91 +65,82 @@ function CodeBasePage(prop) {
...
-          alert(e.response?.status)
-          console.error(e)
-        })
+  const sendPVSBackendRequest = async (method, url) => {
+    const baseURL = 'http://localhost:9100/pvs-api'
+    const requestConfig = {
+      baseURL,
+      url,
+      method,
+      config
```
Comment (Xanonymous-GitHub) [inline src/App/component/CodeBasePage.jsx]:
```suggestion
      headers
```

File: src/App/component/CodeBasePage.jsx [hunk 20]
```diff
@@ -64,91 +65,82 @@ function CodeBasePage(prop) {
...
-        .then(() => {
-          getGitLabCommitFromDB()
-          setLoading(false)
-        })
-        .catch((e) => {
-          alert(e.response?.status)
-          console.error(e)
-        })
+
+    const repo = githubRepo !== undefined ? githubRepo : gitlabRepo
```
Comment (Xanonymous-GitHub) [inline src/App/component/CodeBasePage.jsx]:
```suggestion
    const repo = githubRepo ?? gitlabRepo
```

File: src/App/component/AddRepositoryDialog.jsx (hunk 13 above)
Comment (Xanonymous-GitHub) [reply src/App/component/AddRepositoryDialog.jsx]: For sending a request body by an `axios` instance, you can put it into the `requestConfig`'s `data` field.
please refer to [this line](https://github.com/axios/axios/blob/4fbf61dc751ad69c7df9977d6edf7cf7ee87a5fb/index.d.ts#L85).

File: src/App/component/ComparisonPage.jsx [hunk 21]
```diff
@@ -220,14 +209,42 @@ function ComparisonPage(prop) {
...
+      dataset.labels.push(month.format("YYYY-MM"))
+      for (const branch in dataset.data) {
+        dataset.data[branch].push(getCommitCountFromSelectedBranch(branch, month))
+      }
+    }
+    setDataForComparisonChart(dataset)
+  }
+
+  const getCommitCountFromSelectedBranch = (branch, month) => {
+    if (branch !== "" && branch == leftBranchSelected) {
```
Comment (Xanonymous-GitHub) [inline src/App/component/ComparisonPage.jsx]:
```suggestion
    if (branch !== "" && branch === leftBranchSelected) {
```

File: src/App/component/ComparisonPage.jsx [hunk 22]
```diff
@@ -220,14 +209,42 @@ function ComparisonPage(prop) {
...
+      return commitListDataLeft.filter(commit => {
+        return moment(commit.committedDate).format("YYYY-MM") === month.format("YYYY-MM")
+      }).length
+    }
+    if (branch !== "" && branch == rightBranchSelected) {
```
Comment (Xanonymous-GitHub) [inline src/App/component/ComparisonPage.jsx]:
```suggestion
    if (branch !== "" && branch === rightBranchSelected) {
```

**Comment by Xanonymous-GitHub** [comment]
Since the PR changes are large and have many similarities, if you want to fix a suggestion, please apply it to all other files as well

File: src/App/component/CodeBasePage.jsx [hunk 23]
```diff
@@ -64,91 +65,78 @@ function CodeBasePage(prop) {
...
-  };
-  const handleToggle = () => {
-    setOpen(!open);
-  };
+  const loadingCommitsEnd = () => {
+    setOpen(false)
+  }
+  const loadingCommitsStart = () => {
+    setOpen(!open)
+  }
```
Comment (VincentETN) [inline src/App/component/CodeBasePage.jsx]: 跟前面一樣。

File: src/App/component/ComparisonPage.jsx [hunk 24]
```diff
@@ -68,9 +68,11 @@ const useStyles = makeStyles((theme) => ({
 function ComparisonPage(prop) {
 
   const classes = useStyles()
+  const { startMonth, endMonth } = prop
+
   const [commitListDataLeft, setCommitListDataLeft] = useState([])
   const [commitListDataRight, setCommitListDataRight] = useState([])
-  const [dataForTeamCommitChart, setDataForTeamCommitChart] = useState({ labels: [], data: {} })
+  const [dataForComparisonChart, setDataForComparisonChart] = useState({ labels: [], data: {} })
```
Comment (VincentETN) [inline src/App/component/ComparisonPage.jsx]:
```suggestion
  const [dataOfComparisonChart, setDataOfComparisonChart] = useState({ labels: [], data: {} })
```

Comment (VincentETN) [reply src/App/component/ComparisonPage.jsx]: 小問題，可以延後處理。

File: src/App/component/ComparisonPage.jsx [hunk 25]
```diff
@@ -126,108 +126,106 @@ function ComparisonPage(prop) {
...
-            setCommitListDataRight(responseData)
-          }
-        })
-        .catch((e) => {
-          alert(e)
-          console.error(e)
-        })
+      try {
+        const response = await sendPVSBackendRequest('GET', `/${repo.type}/commits`, {repoOwner, repoName, branchName})
+        whichBranch === "left" ? setCommitListDataLeft(response) : setCommitListDataRight(response)
```
Comment (VincentETN) [inline src/App/component/ComparisonPage.jsx]: 建議別在 get function 裡面 call set dunction。請參考以下說明。
https://github.com/SE-8-2021/frontend/issues/39#issue-1095393925

Comment (hellenGUAN) [reply src/App/component/ComparisonPage.jsx]: 那我把名字改為 setDataOfComparisonChart？

Comment (VincentETN) [reply src/App/component/ComparisonPage.jsx]: 要改名也可以但實際上這 function 現在做了兩件事，名字不好取還要加註解，而且 `setDataOfComparisonChart()` 已經是 useState 的 function。

File: src/App/component/AddRepositoryDialog.jsx [hunk 26]
```diff
@@ -28,82 +28,53 @@ export default function AddRepositoryDialog({ open, reloadProjects, handleClose,
...
-              })
-              .catch((e) => {
-                alert(e.response?.status)
-                console.error(e)
-              })
-          }
-        }).catch((e) => {
-          alert(e.response?.status)
+      const urlCheckResult = await checkRepositoryURL(repositoryURL)
+      if (urlCheckResult === true) {
```
Comment (Xanonymous-GitHub) [inline src/App/component/AddRepositoryDialog.jsx]:
```suggestion
      const isValidUrl = await checkRepositoryURL(repositoryURL)
      if (isValidUrl) {
```

File: src/App/component/AddRepositoryDialog.jsx [hunk 27]
```diff
@@ -28,82 +28,53 @@ export default function AddRepositoryDialog({ open, reloadProjects, handleClose,
...
+        let payload = {
```
Comment (Xanonymous-GitHub) [inline src/App/component/AddRepositoryDialog.jsx]:
```suggestion
        const payload = {
```

File: src/App/component/CodeBasePage.jsx [hunk 28]
```diff
@@ -169,36 +155,43 @@ function CodeBasePage(prop) {
     if (isLoading) {
       const githubRepo = currentProject.repositoryDTOList.find(repo => repo.type === 'github')
       const gitlabRepo = currentProject.repositoryDTOList.find(repo => repo.type === 'gitlab')
-      if (githubRepo !== undefined) {
-        getCommitFromGitHub()
-      }
-      if (gitlabRepo !== undefined) {
-        getCommitFromGitLab()
+      const repo = githubRepo !== undefined ? githubRepo : gitlabRepo
```
Comment (Xanonymous-GitHub) [inline src/App/component/CodeBasePage.jsx]:
```suggestion
      const repo = githubRepo ?? gitlabRepo
```

File: src/App/component/CommitsPage.jsx [hunk 29]
```diff
@@ -64,108 +65,92 @@ function CommitsPage(prop) {
...
-        .then(() => {
-          getGitLabCommitFromDB()
-          setLoading(false)
-        })
-        .catch((e) => {
-          alert(e.response?.status)
-          console.error(e)
-        })
+
+    const repo = githubRepo !== undefined ? githubRepo : gitlabRepo
```
Comment (Xanonymous-GitHub) [inline src/App/component/CommitsPage.jsx]:
```suggestion
    const repo = githubRepo ?? gitlabRepo
```

File: src/App/component/CommitsPage.jsx [hunk 30]
```diff
@@ -174,58 +159,56 @@ function CommitsPage(prop) {
     if (isLoading) {
       const githubRepo = currentProject.repositoryDTOList.find(repo => repo.type === 'github')
       const gitlabRepo = currentProject.repositoryDTOList.find(repo => repo.type === 'gitlab')
-      if (githubRepo !== undefined) {
-        getCommitFromGitHub()
-      }
-      if (gitlabRepo !== undefined) {
-        getCommitFromGitLab()
+      const repo = githubRepo !== undefined ? githubRepo : gitlabRepo
```
Comment (Xanonymous-GitHub) [inline src/App/component/CommitsPage.jsx]:
```suggestion
      const repo = githubRepo ?? gitlabRepo
```

File: src/App/component/ComparisonPage.jsx [hunk 31]
```diff
@@ -126,108 +126,106 @@ function ComparisonPage(prop) {
...
     setRightBranchSelected(e)
-    getCommitFromDBRight(e)
+    getCommitFromDB("right", e)
     setSelectedBranchList([leftBranchSelected, rightBranchSelected])
   }
 
   const handleClick = () => setLoading(true);
 
+  const setComparisonChart = () => {
+    let dataset = { labels: [], data: {} }
```
Comment (Xanonymous-GitHub) [inline src/App/component/ComparisonPage.jsx]:
```suggestion
    const dataset = { labels: [], data: {} }
```

File: src/App/component/ContributionPage.jsx [hunk 32]
```diff
@@ -169,58 +156,59 @@ function ContributionPage(prop) {
     if (isLoading) {
       const githubRepo = currentProject.repositoryDTOList.find(repo => repo.type === 'github')
       const gitlabRepo = currentProject.repositoryDTOList.find(repo => repo.type === 'gitlab')
-      if (githubRepo !== undefined) {
-        getCommitFromGitHub()
-      }
-      if (gitlabRepo !== undefined) {
-        getCommitFromGitLab()
+      const repo = githubRepo !== undefined ? githubRepo : gitlabRepo
```
Comment (Xanonymous-GitHub) [inline src/App/component/ContributionPage.jsx]:
```suggestion
      const repo = githubRepo ?? gitlabRepo
```

File: src/App/component/ContributionPage.jsx [hunk 33]
```diff
@@ -169,58 +156,59 @@ function ContributionPage(prop) {
...
-      labels: [],
-      data: {}
-    }
-    let chartDataset_Deletion = {
-      labels: [],
-      data: {}
-    }
+  const generateBarChart = () => {
+    let chartDataset_Addition = { labels: [], data: {} }
+    let chartDataset_Deletion = { labels: [], data: {} }
```
Comment (Xanonymous-GitHub) [inline src/App/component/ContributionPage.jsx]:
```suggestion
    const chartDataset_Addition = { labels: [], data: {} }
    const chartDataset_Deletion = { labels: [], data: {} }
```

File: src/App/component/AddRepositoryDialog.jsx (hunk 14 above)
Comment (Xanonymous-GitHub) [reply src/App/component/AddRepositoryDialog.jsx]: Hope the next team can improve this comment XDD.

File: src/App/component/AddRepositoryDialog.jsx (hunk 15 above)
Comment (Xanonymous-GitHub) [reply src/App/component/AddRepositoryDialog.jsx]: Hope the next team can improve this comment XDD.

**Review by Xanonymous-GitHub — DISMISSED** [review]

**Review by VincentETN — APPROVED** [review]

**Review by Xanonymous-GitHub — APPROVED** [review]

**Pull request merged**